File: AndroidStudy/GoodWeather/app/src/main/java/com/example/goodweather/bean/dao/AppDatabase.java
package com.example.goodweather.bean.dao;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import com.example.goodweather.bean.Province;

/**
 * @auther zj970
 * @create 2023-06-03 上午12:29
 */
@Database(entities = {Province.class},version = 1,exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "GoodWeatherNew";
    private static volatile AppDatabase mInstance;

    public abstract ProvinceDao provinceDao();

    /**
     * 单例模式
     */
    public static AppDatabase getInstance(Context context) {
        if (mInstance == null) {
            synchronized (AppDatabase.class) {
                if (mInstance == null) {
                    mInstance = Room.databaseBuilder(context.getApplicationContext(),
                            AppDatabase.class, DATABASE_NAME).build();
                }
            }
        }
        return mInstance;
    }
}

File: AndroidStudy/GoodNews/app/src/main/java/com/zj970/goodnews/database/AppDatabase.kt
package com.zj970.goodnews.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.zj970.goodnews.database.bean.ListItem
import com.zj970.goodnews.database.dao.ListItemDao

@Database(entities = [ListItem::class], version = 1, exportSchema = false)
abstract class AppDatabase : RoomDatabase(){
    abstract fun newsItemDao() : ListItemDao
    companion object{
        @Volatile
        private var instance:AppDatabase? = null
        private const val DATABASE_NAME = "good_news.db"
        fun getInstance(context: Context) : AppDatabase{
            return instance ?: synchronized(this){
                instance ?: Room.databaseBuilder(context ,AppDatabase::class.java, DATABASE_NAME)
                    //.allowMainThreadQueries()避免了从主线程中访问数据库 DataStore 本地缓存
                    .build()
                    .also { instance = it }
            }
        }
    }
}